src/contracts/GPv2VaultRelayer.sol

```diff
@@ -44,4 +44,91 @@ contract GPv2VaultRelayer {
     {
         vault.transferFromAccounts(msg.sender, transfers);
     }
+
+    /// @dev Performs a Balancer batched swap on behalf of a user and sends a
+    /// fee to the caller.
+    ///
+    /// This function reverts if:
+    /// - The caller is not the creator of the vault relayer
+    /// - The swap fails
+    /// - The fee transfer fails
+    ///
+    /// @param kind The Balancer swap kind, this can either be `GIVEN_IN` for
+    /// sell orders or `GIVEN_OUT` for buy orders.
+    /// @param swaps The swaps to perform.
+    /// @param tokens The tokens for the swaps. Swaps encode to and from tokens
+    /// as indices into this array.
+    /// @param funds The fund management settings, specifying the user the swap
+    /// is being performed for as well as the recipient of the proceeds.
+    /// @param limits Swap limits for encoding limit prices.
+    /// @param deadline The deadline for the swap.
+    /// @param feeTransfer The transfer data for the caller fee.
+    /// @return tokenDeltas The executed swap amounts.
+    function batchSwapWithFee(
+        IVault.SwapKind kind,
+        IVault.SwapRequest[] calldata swaps,
+        IERC20[] memory tokens,
+        IVault.FundManagement memory funds,
+        int256[] memory limits,
+        uint256 deadline,
+        GPv2Transfer.Data calldata feeTransfer
+    ) external onlyCreator returns (int256[] memory tokenDeltas) {
+        if (kind == IVault.SwapKind.GIVEN_IN) {
+            tokenDeltas = vault.batchSwapGivenIn(
+                swapRequestToIn(swaps),
+                tokens,
+                funds,
+                limits,
+                deadline
+            );
+        } else {
+            tokenDeltas = vault.batchSwapGivenOut(
+                swapRequestToOut(swaps),
+                tokens,
+                funds,
+                limits,
+                deadline
+            );
+        }
+
+        vault.transferFromAccount(msg.sender, feeTransfer);
```

Comment:
Why withdrawing the fees in `batchSwapWithFee` and not directly in the settlement fast path?

Reply:
This saves an extra `GPv2Settlement -> GPv2VaultRelayer` `CALL`, which is a non-negligible amount of gas.

Reply:
naming nit: this reads slightly awkward to me: `transferFromAccount(msg.sender, ..)` could mean you are transferring something from msg.sender.

Reply:
I changed the order of params so it reads more naturally.